classes/form/interval.php

```diff
@@ -0,0 +1,154 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Interval form element
+ *
+ * Contains class to create length of time for element.
+ *
+ * @package local_recompletion
+ * @author Duran CHEN <duran.chen@outlook.com>
+ * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ *
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+global $CFG;
+require_once($CFG->libdir . '/form/group.php');
+require_once($CFG->libdir . '/form/duration.php');
+require_once($CFG->libdir . '/formslib.php');
+require_once($CFG->libdir . '/form/text.php');
+
+class MoodleQuickForm_interval extends MoodleQuickForm_duration {
+
+    /**
+     * Returns time associative array of unit length.
+     *
+     * @return array unit length in seconds => string unit name.
+     */
+    public function get_units() {
+
+        return [
+                'Y' => get_string('years'),
+                'M' => get_string('months'),
+                'D' => get_string('days'),
+                'W' => get_string('weeks'),
+                'TH' => get_string('hours'),
```

Comment:
initially I was unsure about adding a custom form element but the way you handle this with php DateTime\DateInterval looks quite nice - although the only reason we had "seconds/minutes/hours" as options was because it helped me with testing. If we are going to implement a new form element for this we should remove the options for (hours/minutes/seconds) - especially as we now have a process that allows users to trigger a manual recompletion which works quite well for testing. The other benefit to this is that it will simplify this base class a bit as you don't have to handle the "T" types.